Python/ceval.c

```diff
@@ -2476,14 +2476,42 @@ _PyEval_GetBuiltinId(_Py_Identifier *name)
 PyObject *
 PyEval_GetLocals(void)
 {
+    // We need to return a borrowed reference here, so some tricks are needed
     PyThreadState *tstate = _PyThreadState_GET();
      _PyInterpreterFrame *current_frame = _PyThreadState_GetFrame(tstate);
     if (current_frame == NULL) {
         _PyErr_SetString(tstate, PyExc_SystemError, "frame does not exist");
         return NULL;
     }
 
-    PyObject *locals = _PyEval_GetFrameLocals();
+    // Be aware that this returns a new reference
+    PyObject *locals = _PyFrame_GetLocals(current_frame);
+
+    if (locals == NULL) {
+        return NULL;
+    }
+
+    if (PyFrameLocalsProxy_Check(locals)) {
+        PyFrameObject *f = _PyFrame_GetFrameObject(current_frame);
+        PyObject *ret = PyDict_New();
```

Comment:
Regardless of whether it's a proxy or a shared snapshot, the new `f_locals_cache` field needs to be checked to see if it is NULL before getting populated again.

Reply:
Sorry which variable do you refer to?

Reply:
`PyObject *ret = PyDict_New()`

When `f_locals_cache` is already set, it should be reused here rather than allocating a new dictionary (otherwise the desired reference borrowing semantics won't be provided, as subsequent `PyEval_GetLocals()` calls on the same frame would invalidate the reference pointers returned by previous calls)

Reply:
Code suggestion added below.

Reply:
Hmm, I don't agree here. I don't think it's our job to make sure the borrowed reference is "always valid", that's why we need new reference. The user should use borrowed reference in their own risk. For all of our C APIs that return a borrowed reference, the user can do something to invalidate the reference, that's just how borrowed references work.

I think the goal here is to make `PyEval_GetLocals()` mimic `locals()` or `PyEval_GetFrameLocals()`, with the only difference of returned reference type. And to avoid confusion, it should mimic `locals()` in 3.13, not 3.12 (or `PyEval_GetLocals()` in 3.12). We changed the behavior of `locals()` in 3.13 so that it returns a snapshot (that will not be updated) of the current frame locals, and I think the proper behavior of `PyEval_GetLocals()` should be the same - the dict it get should not be updated either, otherwise it will just confuse users.

Reply:
And I really think we should confirm the behavior before updating any documentation. I don't believe making `PyEval_GetLocals()` an exception that is different than `locals()` is the way to go. It should be as simple as - it performs exactly like `locals()` and `PyEval_GetFrameLocals()`, except for the returned reference type. And the original goal is to deprecate it, along with `PyEval_GetGlobals()` and `PyEval_GetBuiltins()`.

Reply:
I know there's a rough implementation in PEP 667, but that implementation does not update the locals cache at all so maybe that's not the bahavior we wanted. If we can't agree on the bahavior, we could involve more people discussing about this. @markshannon maybe as the original author?

Reply:
The main thing I want to ensure is that the following sequence of calls doesn't pose a risk of segfaulting the interpreter:

```
    PyObject *ref1 = PyEval_GetLocals();
    PyObject *ref2 = PyEval_GetLocals();  // This would likely be in a called function rather than inline
    // Do something with ref1 (that should have been safe, since the object the reference is borrowed from is still alive)
```

However, I realised there's a reasonably straightforward way to satisfy all of the following objectives:

* make every call to `PyEval_GetLocals` return a new snapshot on optimised frames
* return a borrowed reference
* keep the borrowed references valid until the frame object goes away

Rather than caching the returned snapshot directly, the snapshots can instead be cached inside a list object. Multiple calls just append new entries to the end of the list. Still thoroughly deprecation worthy, but the segfault risk is gone, while still updating the semantics to match the 3.13 `locals()` behaviour.

(We could optionally delay creating the intermediate list until the second call to `PyEval_GetLocals()` on the same frame, but I'm not sure the memory saving would be big enough to be worth the extra code complexity).

(as far as the docs go, even though they're likely to need further updates, what they say now is closer to reality than what they said after I updated them based on the PEP text that said `PyEval_GetLocals` would start returning a write-through proxy instance. I won't tinker further until this is resolved, though)

Reply:
This was my original thought, but I don't favor that idea now after I talked with Mark (and I still think we should get his input).

I think the key argument is, do we need to ensure this:
```C
    PyObject *ref1 = PyEval_GetLocals();
    PyObject *ref2 = PyEval_GetLocals();  // This would likely be in a called function rather than inline
    // Do something with ref1 (that should be safe, since the frame object the reference is borrowed from is still alive)
```

Like I said, it's trivial for user to crash CPython at C level. The user is responsible to make sure the "borrowed reference" is alive. If the user wants to make sure, the should `Py_INCREF(ref1)`, that's the only reasonable thing to do.

For example,
```C
PyObject* item = PyDict_GetItem(dct, key);
PyDict_SetItem(dct, key, new_value);
// Should we make sure item is alive here?
```

Notice in the example above, the `dct` object is not freed, just like the `FrameObject` in the `PyEval_GetLocals()` case. There's no obligation for us to make sure all the objects "calculated" from object A should live until object A is released.

borrowed reference could be invalidated in many scenarios and users should proceed with precaution. I think that's why we want to replace our borrowed reference APIs with new reference. I don't think that we should do the list here just to prevent users from crashing themselves with dangerous code.

The cost is not huge, I agree, but the memory will be used until the frame is released (if user uses `PyEval_GetLocals()` in a loop inside a function? That's `O(N)` increment for memory when the user thinks they are using borrowed reference).

That's my thoughts about this issue. Again I might not be correct, and I'm open to other suggestions and opinions.

Reply:
The thing with borrowed references is that they're borrowed *from* another object, and in the case of `PyEval_GetLocals()` that object is the frame object. In previous versions, as long as the frame object was still alive, then the borrowed locals reference remained valid, including in the presence of multiple calls to `PyEval_GetLocals()`.

If we don't preserve that property, then C code that worked fine on Python 3.12 will likely start segfaulting on Python 3.13. If PEP 667 had said that was going to happen, then I expect the SC would have made a request to figure out a way to avoid the segfault risk. (Neither the text nor the example code in PEP 667 carried a segfault risk, they just differed in whether the cached value was a proxy creating a reference loop or a snapshot dict that no other APIs would access)

Don't get me wrong, I don't think using a list of references is a *good* option, I just think it's the *least bad* one if we want to convert the function's semantics over to those of the Python 3.13 `locals` builtin.

I'm also fine with leaving it back on the Python 3.12 `locals` builtin semantics, where repeated calls refresh a shared snapshot (I actually prefer that option, since existing users of the API are necessarily familiar with that behaviour, it doesn't increase memory usage relative to Python 3.12, and new users should be using `PyEval_GetFrameLocals()` anyway, so if the weird legacy semantics push them away from `PyEval_GetLocals()` that's a good thing). 

Of all the APIs deprecated in PEP 667, `PyEval_GetLocals()` is the only one I'd like to see upgraded to a hard deprecation such that it eventually goes away (while the SC already said "no" to a hard deprecation in 3.13, we'll hopefully be able to reconsider that in 3.14).

Reply:
I understand your point. My argument is that it would be a very rare case where the user updates the cache with `PyEval_GetLocals()` and still try to access the previous one. The list solution is so bad (and it's not your fault it's just what we have to do to keep `locals()` behavior + save all possible segfaults) that I would also lean to keeping the 3.12 behavior just to make the code cleaner.

I still think we need more eyes on this to reach a conclusion. Our possible solutions are all very clear and we both know the pros and cons for each, but we have different opinions and evaluations. Maybe involving more people, even SC in this decision, could solve the problem for good. I won't be persistent if most of the people think the alternatives are better.

Reply:
I don't think anyone would be directly calling PyEval_GetLocals multiple times in the same function, but I *do* consider it entirely plausible that they may be calling other APIs in their application or extension module that also call PyEval_GetLocals independently.

So I think our disagreement boils down to two points:

* I think the above scenario, while genuinely unlikely, is plausible enough that we shouldn't risk making it segfault. You consider it so unlikely that it isn't really worth worrying about.
* You see more value in making PyEval_GetLocals consistent with the semantics of the rest of Python 3.13 than I do

I'll summarise this on the open SC ticket, and mention that ticket in the release discussion channel in Discord. 

(Originally I thought caching and returning a proxy instance offered the best of both worlds, but changed my mind when I realised it posed a different set of segfault risks if anyone had assumed it was a dict without properly checking the runtime type)

Reply:
For the SC question, I'll present all three options:

* keep Python 3.12 semantics
* adopt Python 3.13 semantics by invalidating previously borrowed references on repeated calls
* adopt Python 3.13 semantics by keeping a list of references

and let them know neither of us actually prefer the third option, but it does exist

Reply:
That is very thorough and represents exactly what I thought. Thank you for putting together the options (I'm not very good with the words and the documentation LOL)

Reply:
Your explanations here have been entirely clear - I'm definitely at a point where I think any of the three options would actually be mostly OK (the only adjustment I'd argue for is that if we did go down the `Py_XSETREF` path we add a hard deprecation warning that gets triggered when `PyEval_GetLocals()` is called on an optimised frame with `f_locals_cache` already set. That way if folks did manage to get a segfault, they'd hopefully see the deprecation warning first)

Summary has been posted here (trying to be as fair to both points of view as I can): https://github.com/python/steering-council/issues/245#issuecomment-2148820321

I'll also ping @Yhg1s as release manager in the Discord release-discussions channel.

Reply:
@markshannon recommended in https://github.com/python/steering-council/issues/245#issuecomment-2155093992), that we trust the original PEP text over its sample code and cache a locals proxy instance instead of a snapshot (i.e. bring `PyEval_GetLocals()` down on the side of aligning with `f.f_locals` access rather than with `locals()`)

I'm certainly fine with that option (going back to those docs mostly involves restoring text I had previously revised), so if @gaogaotiantian is also OK with it, there's no longer a disagreement for the SC to resolve.